### PR TITLE
python3Packages.jax: 0.2.28 -> 0.3.0

### PR DESCRIPTION
**Pull request by samuela**
###### Motivation for this change
Upgrades python3Packages.jax from 0.2.28 to 0.3.0

This PR was automatically generated by [nixpkgs-upkeep](https://github.com/samuela/nixpkgs-upkeep).
- [CI workflow](https://github.com/samuela/nixpkgs-upkeep/actions/runs/1826793821) that created this PR.
- Internal tag: 42ab84d8819d796e64f8234cb79e6347.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuela** [comment]
Currently blocked on https://github.com/cgarciae/treeo/issues/7, but maybe we should just patch around that.

jaxlib (https://github.com/google/jax/releases/tag/jaxlib-v0.3.0) will also need to be upgraded manually

**Comment by samuela** [comment]
currently blocked on https://github.com/google/jax/issues/9534

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 159099` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python39Packages.elegy</li>
  </ul>
</details>
<details>
  <summary>18 packages built:</summary>
  <ul>
    <li>python310Packages.chex</li>
    <li>python310Packages.jax</li>
    <li>python310Packages.jaxlib (python310Packages.jaxlib-build ,python310Packages.jaxlibWithoutCuda)</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.jmp</li>
    <li>python310Packages.treeo</li>
    <li>python39Packages.chex</li>
    <li>python39Packages.dm-haiku</li>
    <li>python39Packages.flax</li>
    <li>python39Packages.jax</li>
    <li>python39Packages.jaxlib (python39Packages.jaxlib-build ,python39Packages.jaxlibWithoutCuda)</li>
    <li>python39Packages.jaxlib-bin</li>
    <li>python39Packages.jaxlibWithCuda</li>
    <li>python39Packages.jmp</li>
    <li>python39Packages.objax</li>
    <li>python39Packages.optax</li>
    <li>python39Packages.treeo</li>
    <li>python39Packages.treex</li>
  </ul>
</details>

**Comment by samuela** [comment]
elegy failure is unrelated:
```
builder for '/nix/store/anxmyb0011g41krphyxb5jidl9rvarn9-python3.9-elegy-0.8.4.drv' failed with exit code 1; last 10 log lines:
  Successfully built elegy
  Finished creating a wheel...
  Finished executing pipBuildPhase
  installing
  Executing pipInstallPhase
  /build/source/dist /build/source
  Processing ./elegy-0.8.4-py3-none-any.whl
  Requirement already satisfied: tensorboardx<3.0,>=2.1 in /nix/store/vfpnjl9hnrlpd5pf90ibp5zavywm7pqi-python3.9-tensorboardx-2.4/lib/python3.9/site-packages (from elegy==0.8.4) (2.4)
  ERROR: Could not find a version that satisfies the requirement cloudpickle<2.0.0,>=1.5.0 (from elegy) (from versions: none)
  ERROR: No matching distribution found for cloudpickle<2.0.0,>=1.5.0
cannot build derivation '/nix/store/2shnhq0f7cxx7gk4f8fjxqasbm82vpql-review-shell.drv': 1 dependencies couldn't be built
```

This PR is ready for review.

**Comment by samuela** [comment]
elegy failure is not tracked: https://github.com/NixOS/nixpkgs/issues/159456

**Comment by samuela** [comment]
absent a review, I'll merge tomorrow unless anyone objects

**Review by ndl — COMMENTED** [review]
Overall SG but pinning protobuf version might potentially be problematic, see the inline comment.

File: pkgs/development/python-modules/jaxlib/default.nix
Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:112]: Ahh that's a great catch... I wonder if this would be problematic at all. It seems that protobuf isn't included in `propagatedBuildInputs` anywhere though, so I don't think they'd conflict in PYTHONPATH... 

I does give me some https://github.com/NixOS/nixpkgs/issues/156767 flashbacks though. 

I didn't try patching since that seemed like a dangerous rabbit hole, but maybe I'll give it a shot.

Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:112]: Ok I just tested this and everything seems to work fine.
```
❯ NIX_PATH=.. nix-shell -p python3Packages.jaxlib python3Packages.tensorflow                                  

[nix-shell:~/dev/nixpkgs]$ echo $PYTHONPATH | tr ':' '\n' | grep protobuf
/nix/store/j7jbqfagpar0h2xvklzxrjp74443fqaf-python3.9-protobuf-3.19.3/lib/python3.9/site-packages

[nix-shell:~/dev/nixpkgs]$ echo $LD_LIBRARY_PATH | tr ':' '\n' | grep protobuf

[nix-shell:~/dev/nixpkgs]$ echo $PATH | tr ':' '\n' | grep protobuf
/nix/store/kyajg85hhf8anb8rk0j5dpg9ma9yyyxk-protobuf-3.19.3/bin

[nix-shell:~/dev/nixpkgs]$ python
Python 3.9.10 (main, Jan 13 2022, 23:32:03) 
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import jaxlib
>>> import tensorflow
>>> 
```
I think because protobuf isn't in `propagatedBuildInputs` in jaxlib.

Does this look ok to you @ndl? Is there anything I might be missing?

Comment (ndl) [reply pkgs/development/python-modules/jaxlib/default.nix:112]: Just to clarify - I'm not concerned about Python `protobuf` versions conflict but about the underlying `.so` libraries conflict. 

The resulting Python `.so` extensions for `jaxlib` contain the reference to `libprotobuf.so`:

```
ldd /nix/store/lqv8zg74ha8zw33kkr4ybh05zd27qiin-python3.9-jaxlib-0.1.75/lib/python3.9/site-packages/jaxlib/xla_extension.so | grep protobuf
```
should show smth like

```
libprotobuf.so.30 => /nix/store/kyajg85hhf8anb8rk0j5dpg9ma9yyyxk-protobuf-3.19.3/lib/libprotobuf.so.30
```

In addition, if you run the following command:
```
LD_DEBUG=files nix-shell -p "with (import ./.{}); python3.withPackages(ps: [ ps.jaxlib ps.jax ])" --run "python3 -c 'import jax.numpy as jnp; jnp.random()'" 2>&1 | grep protobuf.so
```
you'll get the output that looks like this:
```
calling init: /nix/store/kyajg85hhf8anb8rk0j5dpg9ma9yyyxk-protobuf-3.19.3/lib/libprotobuf.so.30
```

This demonstrates that importing and using `jax` (which, in turn, will import and use `jaxlib`) will cause `libprotobuf.so` to be loaded into the process.

Now, if there's another package that also uses `libprotobuf.so` (either Python `protobuf` implementation or any other package that links to `libprotobuf.so`), which `libprotobuf.so` gets imported depends on the import order of the packages. If they all import the same `libprotobuf.so` - that doesn't matter, but if the versions are different - you might get some "interesting" (and somewhat hard to debug) issues.

Now, I don't want to block your PR based on this - fully hypothetical ATM - scenario, just explaining what I meant in my comments above and what types of problematic scenarios _might_ surface :-)

Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:112]: Ah ok, thanks for explaining! My mental model was that things would be ok as long as each .so doesn't link against two versions of protobuf, causing a conflict in the symbol table (ie everything's ok as long as ldd only has one protobuf). But is this still problematic if A.so links against protobuf1 and B.so links against protobuf2?

Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:112]: I guess these are the sorts of cases where dynamic linking is so annoying...

Comment (ndl) [reply pkgs/development/python-modules/jaxlib/default.nix:112]: > is this still problematic if A.so links against protobuf1 and B.so links against protobuf2?

In general - "it depends" :-) I'm by no means an expert in this topic, but IIUC depending on whether you use dynamic loading via `dlopen` [with some fancy flags](https://stackoverflow.com/questions/36545396/dlopen-doesnt-respect-rtld-local), play with [symbols visibility](https://stackoverflow.com/questions/435352/limiting-visibility-of-symbols-when-linking-shared-libraries) and / or use [symbols versioning](https://man7.org/conf/lca2006/shared_libraries/slide19a.html), use different [SONAME](https://en.wikipedia.org/wiki/Soname), etc - you can get all sorts of interesting combinations and results.

Speaking of our particular case: `protobuf` [does seem to use SONAME for versioning](https://github.com/protocolbuffers/protobuf/blob/master/src/Makefile.am#L21), so different dynamic libraries versions will get different library names and will be loaded separately. It also uses [visibility control](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/port_def.inc#L437) so, in principle, it should export only its APIs (= reducing the surface for symbols conflicts). OTOH it doesn't look like there's any symbols versioning policy, so I'd still expect the conflicts of symbols are possible when two different versions of the library are loaded. I'm not 100% sure if the errors [like this](https://stackoverflow.com/questions/56801994/how-to-use-shared-library-with-two-different-versions-indirectly-in-my-project) are related to symbols conflicts or smth else, though - needs deeper investigation.

... I guess that's a long and convoluted way of saying "I still don't know whether it will work" :-) Simple test case of importing two different Python extensions linked with different protobuf shared libraries versions "works" (= doesn't crash), but I also haven't tried doing anything meaningful there.

I guess that if patching the code or switching to static linking is not an option, you can try going ahead and using your current PR "as is" and see if it causes any issues in practice?

Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix:112]: Well you sure sound like an expert to me 👍  At least I'm learning a lot here, which is great! Thank you so much for your detailed response!

Let me take a stab at patching. If that proves to be sufficiently complicated I vote that we merge this as is and try to fix in a future PR. Hopefully the SONAME versioning is sufficient to avoid conflicts. At minimum I think we should be compatible with `python3Packages.tensorflow`. From what I understand there's no way for us to eliminate the possibility of symbol collisions entirely since users could always import some arbitrary other python package that happens to depend load an so with a conflicting protobuf version anyway. So the best we can do is to try to be good nixpkgs citizens and follow the latest `protobuf`.

Comment (samuela) [reply pkgs/development/python-modules/jaxlib/default.nix]: Implemented the patch solution in https://github.com/NixOS/nixpkgs/pull/159099/commits/f3753388ad7291ee20f9bd07c25b9a66ed62269a. Not as messy as I expected!

**Comment by samuela** [comment]
With https://github.com/NixOS/nixpkgs/pull/159099/commits/f3753388ad7291ee20f9bd07c25b9a66ed62269a we are now building against `protobuf` instead of `protobuf3_12`. An interesting note about that commit: `fetchAttrs` hashes had to be updated for the CPU version, but not for the GPU version for some reason. Either way, both versions build just fine now.

I'll go ahead and squash and merge tomorrow unless anyone objects!

File: pkgs/development/python-modules/treeo/default.nix
Comment (ndl) [inline pkgs/development/python-modules/treeo/default.nix:3]: Why is `fetchpatch` needed? It doesn't seem to be used?

File: pkgs/development/python-modules/treeo/default.nix
Comment (ndl) [inline pkgs/development/python-modules/treeo/default.nix:12]: Given https://github.com/NixOS/nixpkgs/pull/159933 do we need this change in this PR?

**Review by ndl — APPROVED** [review]
Great, thanks for spending the time on fixing the `protobuf` versioning situation!

I have just one extra question RE: `treeo` change (sorry, didn't notice it before - we might want to skip it given https://github.com/NixOS/nixpkgs/pull/159933) but otherwise LGTM.

File: pkgs/development/python-modules/treeo/default.nix
Comment (samuela) [reply pkgs/development/python-modules/treeo/default.nix:12]: we can skip it in this pr if you'd prefer. I just wanted to make everything go green, but I'm happy if https://github.com/NixOS/nixpkgs/pull/159933 looks good to you

Comment (ndl) [reply pkgs/development/python-modules/treeo/default.nix:12]: Just left a comment on https://github.com/NixOS/nixpkgs/pull/159933, I suggest we unbundle `treeo` versioning issue and deal with it in that separate PR so that you can go ahead and merge this one?

**Comment by samuela** [comment]
@ndl ok just force pushed new commits to upgrade jax, jaxlib, jaxlib-bin but avoid touching treeo. Do those commits look ok to you?

**Comment by ndl** [comment]
@samuela yes, the current set of changes LGTM, thank you!

**Comment by samuela** [comment]
ok great, thanks for all your help @ndl! merging now...

**Pull request merged**
